test/jdk/java/foreign/TestMappedHandshake.java

```diff
@@ -0,0 +1,142 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @requires vm.flavor != "zero"
+ * @modules java.base/jdk.internal.vm.annotation java.base/jdk.internal.misc
+ * @key randomness
+ * @run testng/othervm TestMappedHandshake
+ * @run testng/othervm -Xint TestMappedHandshake
+ * @run testng/othervm -XX:TieredStopAtLevel=1 TestMappedHandshake
+ * @run testng/othervm -XX:-TieredCompilation TestMappedHandshake
+ */
+
+import java.io.File;
+import java.io.IOException;
+import java.lang.foreign.Arena;
+import java.lang.foreign.MemorySegment;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.testng.annotations.Test;
+
+import static org.testng.Assert.*;
+
+public class TestMappedHandshake {
+
+    static final int SEGMENT_SIZE = 1_000_000;
+    static final int ACCESS_START_DELAY_MILLIS = 100;
+    static final int POST_ACCESS_DELAY_MILLIS = 1;
+    static final int TIMED_RUN_TIME_SECONDS = 10;
+    static final int MAX_EXECUTOR_WAIT_SECONDS = 20;
+
+    static final int NUM_ACCESSORS = 5;
+
+    static final Path tempPath;
+
+    static {
+        try {
+            File file = File.createTempFile("buffer", "txt");
+            file.deleteOnExit();
+            tempPath = file.toPath();
+            Files.write(file.toPath(), new byte[SEGMENT_SIZE], StandardOpenOption.WRITE);
+
+        } catch (IOException ex) {
+            throw new ExceptionInInitializerError(ex);
+        }
+    }
+
+    @Test
+    public void testHandshake() throws InterruptedException, IOException {
+        try (FileChannel fileChannel = FileChannel.open(tempPath, StandardOpenOption.READ, StandardOpenOption.WRITE) ;
+             Arena arena = Arena.ofShared()) {
+            MemorySegment segment = fileChannel.map(FileChannel.MapMode.READ_WRITE, 0L, SEGMENT_SIZE, arena);
+            ExecutorService accessExecutor = Executors.newFixedThreadPool(NUM_ACCESSORS + 1);
+            // start handshaker
+            accessExecutor.execute(new Handshaker());
+            Thread.sleep(ACCESS_START_DELAY_MILLIS);
+            // start accessors
+            for (int i = 0 ; i < NUM_ACCESSORS ; i++) {
+                accessExecutor.execute(new MappedSegmentAccessor(segment));
+            }
+
+            accessExecutor.shutdown();
+            assertTrue(accessExecutor.awaitTermination(MAX_EXECUTOR_WAIT_SECONDS, TimeUnit.SECONDS));
```

Comment:
ExecutorService is an AutoCloseable to you should be able to use try-with-resources and get rid of the the shutdown+awaitTermination.

Reply:
The test (like also its sibling `TestHandshake`) is deliberately using `shutdown`/`awaitTermination` to make sure the test finished in a reasonable amount of time (and generate correct assertion failures if it doesn't). This was mostly added to have better debugging when `TestHandshake` was failing spuriously, so I'd prefer to leave it as is.